mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeans.scala

```diff
@@ -0,0 +1,690 @@
+    @Since("1.6.0") val center: Vector,
+    @Since("1.6.0") val rows: Long,
+    @Since("1.6.0") val criterion: Double,
+    private var localHeight: Double,
+    private var parent: Option[BisectingClusterNode],
+    private var children: Seq[BisectingClusterNode]) extends Serializable {
+
+  require(!criterion.isNaN)
+
+  @Since("1.6.0")
+  def this(center: Vector, rows: Long, criterion: Double) =
+    this(center, rows, criterion, 0.0, None, Array.empty[BisectingClusterNode])
+
+  /**
+   * Inserts a sub node as its child
+   *
+   * @param child inserted sub node
+   */
+  @Since("1.6.0")
+  def insert(child: BisectingClusterNode) {
+    insert(Array(child))
+  }
+
+  /**
+   * Inserts sub nodes as its children
+   *
+   * @param children inserted sub nodes
+   */
+  @Since("1.6.0")
+  def insert(children: Array[BisectingClusterNode]) {
+    this.children = this.children ++ children
+    children.foreach(child => child.parent = Some(this))
+  }
+
+  /**
+   * Converts the tree into Array class
+   * the sub nodes are recursively expanded
+   *
+   * @return an Array class which the cluster tree is expanded
+   */
+  @Since("1.6.0")
+  def toArray: Array[BisectingClusterNode] = {
+    val array = this.children.size match {
+      case 0 => Array(this)
+      case _ => Array(this) ++ this.children.flatMap(child => child.toArray.toIterator)
+    }
+    array.sortWith { case (a, b) =>
+      a.getDepth < b.getDepth && a.criterion < b.criterion && a.rows < b.rows
+    }
+  }
+
+  /**
+   * Gets the depth of the cluster in the tree
+   *
+   * @return the depth from the root
+   */
+  @Since("1.6.0")
+  def getDepth: Int = {
+    this.parent match {
+      case None => 0
+      case _ => 1 + this.parent.get.getDepth
+    }
+  }
+
+  /**
+   * Gets the leaves nodes in the cluster tree
+   */
+  @Since("1.6.0")
+  def getLeavesNodes: Array[BisectingClusterNode] = {
+    this.toArray.filter(_.isLeaf).sortBy(_.center.toArray.sum)
+  }
+
+  @Since("1.6.0")
+  def isLeaf: Boolean = this.children.isEmpty
+
+  @Since("1.6.0")
+  def getParent: Option[BisectingClusterNode] = this.parent
+
+  @Since("1.6.0")
+  def getChildren: Seq[BisectingClusterNode] = this.children
+
+  /**
+   * Gets the dendrogram height of the cluster at the cluster tree.
+   * A dendrogram height is different from a local height.
+   * A dendrogram height means a total height of a node in a tree.
+   * A local height means a maximum distance between a node and its children.
+   *
+   * @return the dendrogram height
+   */
+  @Since("1.6.0")
+  def getHeight: Double = {
+    this.children.size match {
+      case 0 => 0.0
+      case _ => this.localHeight + this.children.map(_.getHeight).max
+    }
+  }
+
+  @Since("1.6.0")
+  def setLocalHeight(height: Double): Unit = this.localHeight = height
+
+  /**
+   * Converts to an adjacency list
+   *
+   * @return List[(fromNodeId, toNodeId, distance)]
+   */
+  @Since("1.6.0")
+  def toAdjacencyList: Array[(Int, Int, Double)] = {
```

Comment:
Maybe this is too much for the first PR. We can remove it and address it in the future.